File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode*find(vector<int>&preorder,vector<int>&inorder,int s,int e,int &index,unordered_map<int,int>&mp)
    {
        if(s>e)
        {
            return NULL;
        }
        TreeNode*root=new TreeNode(preorder[index]);
         int i=mp[preorder[index]];
         index++;
         root->left=find(preorder,inorder,s,i-1,index,mp);
         root->right=find(preorder,inorder,i+1,e,index,mp);
         return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int,int>mp;
        for(int i=0;i<inorder.size();i++)
        {
            mp[inorder[i]]=i;
        }
        int index=0;
         return find(preorder,inorder,0,inorder.size()-1,index,mp);
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int index=0;
    public  TreeNode find(int[]preorder,int[]inorder,int s,int e,HashMap<Integer,Integer>map)
    {
        if(s>e)
        {
            return null;
        }
        int rootVal=preorder[index];
        index++;
        TreeNode root=new TreeNode(rootVal);
        int i=map.get(rootVal);
        root.left=find(preorder,inorder,s,i-1,map);
        root.right=find(preorder,inorder,i+1,e,map);
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer,Integer>map=new HashMap<>();
        int n=inorder.length;
        for(int i=0;i<n;i++)
        {
            map.put(inorder[i],i);
        }
        return find(preorder,inorder,0,n-1,map);
    }
}